### DISCUSSION TITLE
What breaks if you omit --no-build-isolation with editable installs?

### DISCUSSION THREAD

**Discussion by inklesspen** [mesonbuild/meson-python#628] [Q&A] [answered]
The doc page on [Editable Installs](https://meson-python.readthedocs.io/en/latest/how-to-guides/editable-installs.html#build-dependencies) says that "packages installed in editable mode are rebuilt on import", and that failure to use `--no-build-isolation` can cause dependencies to be missing at import time.

However, in my experience running `pip install -e .`, the build directory persists even without this flag. Importing the package works just fine, and in fact I can't see any difference between running it with that flag or without it (except of course that I have to manually install the build dependencies when using that flag).

Here's my sample project: https://github.com/inklesspen/meson-python-cffi-example

Perhaps the breakage occurs with features I've not yet used? Or perhaps this is intended for versions of pip [older than 21.3](https://pip.pypa.io/en/stable/topics/local-project-installs/#build-artifacts)?

**Comment by eli-schwartz** [comment]
The issue is that you have to manually install build dependencies because the ones originally installed during the build got uninstalled.

Also: they may be different versions which means that you may need to reconfigure the build anyway. Worse dependencies may have been originally detected and cached in a location that doesn't exist and doesn't work anymore, resulting in the build failing.

**Reply by inklesspen** [reply]
Okay, but if you have an editable install you're hacking on the project anyway, right? So if those dependencies get moved you're going to have to resolve that anyway.

**Reply by eli-schwartz** [reply]
Which is why the recommendation is to not use temporary copies of the dependencies, so they don't get moved. :)

**Reply by rgommers** [reply]
I had a look at the sample project. It does have a build-time dependency (`cffi`) that will go missing without `--no-build-isolation` if it's not already installed in the base environment. The failure mode will show up after making an edit to `_fontconfig.c.py` and then trying to use the editable install.

@inklesspen I think the first answer from @eli-schwartz here is correct. If it's clear, could you please mark that as the answer?

**Reply by inklesspen** [reply]
i guess? i don't understand why/how that dependency would go missing without something happening to break the whole python environment, but i suppose it is always possible to deliberately shoot yourself in the foot.

**Reply by eli-schwartz** [reply]
The something that happens is "pip".

Pip itself is defined to break the whole python environment. Using pip install -e without --no-build-isolation is, well, "deliberately shooting yourself in the foot" except that actually it is the PyPA standards shooting you in the foot.

It's a bad design, absolutely, but it's also very much absolutely the design that exists.

Build isolation and editable installs were designed ad-hoc, at different times, by different people. Then the latter was hotfixed by people who work with pure `*.py` projects and don't compile C, not even FFI. The way they work is very inconsistent.

Build isolation means that build dependencies are *installed into a virtualenv and then deleted via rm -rf* as soon as pip completes. Editable installs, meanwhile, mean that the build directory is kept intact (but not dependencies, since editable installs have no power over the build isolation virtualenv).

This is "the whole point" of build isolation, apparently. That's why you need to disable it for editable installs.

**Reply by rgommers** [reply]
>  i don't understand why/how that dependency would go missing without something happening to break the whole python environment

To be clearer, this is what happens:

1. `pip` creates a temporary virtual environment (let's call it `bld_venv`)
2. `pip` activates `bld_venv` and installs the build dependencies listed in `pyproject.toml` into `bld_venv`
3. `pip` then asks the build backend (`meson-python` in this case) to build the package
4. `pip` then deletes `bld_venv` (here your `cffi` goes missing)

Then, when you use your editable install, `ninja` will run again to rebuild the package incrementally - and if a source file changed something does need rebuilding. But `bld_venv` is gone, so that does not work.

**Reply by inklesspen** [reply]
But that doesn't happen…

1. I create a virtual environment and activate it. (I use direnv for that, but I could probably do it manually too.)
2. I run `pip install -e .`
3. pip installs cffi and meson-python in the activated virtualenv, then builds and installs my package
4. pip does not, in fact, delete the activated virtualenv, nor the build directory
5. I run python and import the package and everything works fine.

**Reply by rgommers** [reply]
`pip` 24.0 definitely does this for me, and it always has AFAIK. Note that the logging won't actually say it creates a venv, even when you add `-v` - but it does. Can you please make sure with an empty venv without `direnv`? If you have a reproducer from scratch where build dependencies stay available, I'll try it.

> nor the build directory

Yes, the build dir is left in place.

**Reply by inklesspen** [reply]
All right, I've [sorted it out](https://gist.github.com/inklesspen/4de53c385d207624bd2766728ce1c6c6#file-readme-md); click that link for Dockerfiles to make a reproducible example.

It turns out that when meson and ninja are installed at the system level (as they are in my regular work machine) you don't see the breakage from an editable install with build isolation; if meson and ninja are _not_ installed system-wide, you do see the breakage (since pip installs them in a temporary directory, as you said).

**Reply by rgommers** [reply, marked as answer]
Thank you, that is indeed reproducible and all makes sense. The one issue in my explanation above is that `cffi` didn't actually go missing, but that is because you also declared it as a runtime dependency with `dependencies = ['cffi']` so it does end up in `myvenv`. You can check that declared build-only deps don't actually get preserved in your `example.working` with:

```
$ docker run --rm -it example:working bash
$ /home/builder/myvenv/bin/python3 -c "import cffi"  # works, because also a runtime dep
$ /home/builder/myvenv/bin/python3 -c "import mesonpy"  # build-only dep, so went missing
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'mesonpy'
```

**Reply by inklesspen** [reply]
I'm pretty sure it _is_ a runtime dependency, so yeah. Anyway, much is explained now. Thanks.
